File: NC/linked-list-cycle-detection.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        if head is None:
            return False

        slow = head
        fast = head.next

        while slow is not fast:
            if fast is None:
                return False
            slow = slow.next
            fast = fast.next
            if slow is None or fast is None:
                return False

            fast = fast.next

        return True

File: NC/linked-list-cycle-detection.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution
{
public:
    bool hasCycle(ListNode *head)
    {
        if (head == NULL)
            return false;

        ListNode *slow = head, *fast = head;

        while (slow && head)
        {
            fast = fast->next;
            if (fast == NULL)
            {
                return false;
            }
            fast = fast->next;
            if (fast == NULL)
            {
                return false;
            }
            slow = slow->next;
            if (fast == slow)
            {
                return true;
            }
        }
        return false;
    }
};
